Source/ProgressionSystemRuntime/Public/Components/PSHUDComponent.h

```diff
@@ -0,0 +1,134 @@
+	UPSHUDComponent();
+
+	/** Returns the Progression System data asset. */
+	UFUNCTION(BlueprintPure, Category = "C++")
+	static const UPSDataAsset* GetProgressionSystemDataAsset() { return &UPSDataAsset::Get(); }
+
+	/** Returns current saved progression. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	const FORCEINLINE FPSRowData& GetSavedProgressionRowData() const { return SavedProgressionRowDataInternal; }
+
+	/** Save the progression depends on EEndGameState. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	void SavePoints(ELevelType Map, FPlayerTag Character, EEndGameState EndGameState);
+
+	/** Returns the endgame reward. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	int32 GetProgressionReward(ELevelType Map, FPlayerTag Character, EEndGameState EndGameState);
+
+	/** Saves the current progression. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	void SaveCurrentGameProgression();
+
+	UFUNCTION()
+	void SaveDataAsync();
+
+	/*********************************************************************************************
+	* Protected properties
+	********************************************************************************************* */
+protected:
+	/** Progression System data asset */
+	UPROPERTY(Transient, BlueprintReadOnly, meta = (BlueprintProtected, DisplayName = "Progression System Data Asset"))
+	TSoftObjectPtr<UPSDataAsset> ProgressionSystemDataAssetInternal;
+
+	/** Created Main Menu widget. */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Transient, Category = "C++", meta = (BlueprintProtected, DisplayName = "Progression Menu Widget"))
+	TObjectPtr<class UPSMenuWidget> ProgressionMenuWidgetInternal = nullptr;
+
+	/** Created Save points widget. */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Transient, Category = "C++", meta = (BlueprintProtected, DisplayName = "Progression Save Widget"))
+	TObjectPtr<class UPSSaveWidget> ProgressionSaveWidgetInternal = nullptr;
+
+	/** Store the save game instance */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Category = "C++", meta = (BlueprintProtected, DisplayName = "Save Game Instance"))
+	TObjectPtr<class UPSSaveGameData> SaveGameInstanceInternal = nullptr;
+
+	/** The current Saved Progression of a player. */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Category = "C++", meta = (BlueprintProtected, DisplayName = "Saved Progression Row Data"))
+	FPSRowData SavedProgressionRowDataInternal = FPSRowData::EmptyData;
+
+	/** The current selected player */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Category = "C++", meta = (BlueprintProtected, DisplayName = "Current Player Tag"))
+	FPlayerTag CurrentPlayerTagInternal = FPlayerTag::None;
+
+	/** The Progression Data Table that is responsible for progression configuration. */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Transient, Category = "C++", meta = (BlueprintProtected, DisplayName = "Progression Data Table"))
+	TObjectPtr<UDataTable> ProgressionDataTableInternal = nullptr;
```

Comment:
Do you really need to cache the data table? All table rows have to be obtained only once on starting the game and cached in some singleton like your Subsystem, then other systems work with cached result instead of referencing Data Table directly